ReSharper.FSharp/src/FSharp/FSharp.Psi.Features/src/AICore/FSharpFileSummarizer.fs

```diff
@@ -1,36 +1,271 @@
+        | :? FSharpMemberOrFunctionOrValue as mfv -> formatMfv mfv displayContext true
+        | _ -> "_"
+
+    static let rec formatTypeUsage (typeUsage: ITypeUsage) =
+        match typeUsage with
+        | :? INamedTypeUsage as typeUsage ->
+            formatTypeReferenceName typeUsage.ReferenceName
+
+        | :? ITupleTypeUsage as typeUsage ->
+            typeUsage.Items |> Seq.map formatTypeUsage |> String.concat " * "
+
+        | :? IParenTypeUsage as typeUsage ->
+            $"({formatTypeUsage typeUsage.InnerTypeUsage})"
+
+        | :? IFunctionTypeUsage as typeUsage ->
+            $"{formatTypeUsage typeUsage.ArgumentTypeUsage} -> {formatTypeUsage typeUsage.ReturnTypeUsage}"
+
+        | :? IArrayTypeUsage as typeUsage ->
+            $"{formatTypeUsage typeUsage.TypeUsage}[]"
+
+        | :? IAnonRecordTypeUsage as typeUsage ->
+           let fields =
+               typeUsage.Fields
+               |> Seq.map (fun field -> $"{field.ReferenceName.ShortName}: {formatTypeUsage field.TypeUsage}")
+               |> String.concat "; "
+           $"{{|{fields}|}}"
+
+        | :? IWithNullTypeUsage as typeUsage ->
+            $"{formatTypeUsage typeUsage.TypeUsage} | null"
+
+        | _ -> ""
+
+    and formatTypeReferenceName (typeName: ITypeReferenceName) =
+        let name = typeName.Identifier.GetTypeParameterOrSourceName()
+        let typeArgs = typeName.TypeArgumentList
+        if isNull typeArgs then name else
+        let typeUsages = typeArgs.TypeUsages
+        if Seq.isEmpty typeUsages then name else
+
+        let typeArgsRepresentation =
+            typeUsages
+            |> Seq.map formatTypeUsage
+            |> String.concat ", "
+
+        match typeArgs with
+        | :? IPrefixAppTypeArgumentList -> $"{name}<{typeArgsRepresentation}>"
+        | _ -> $"{typeArgsRepresentation} {name}"
+
+    let formatTypeOrExtDeclaration (typeDecl: IFSharpTypeOrExtensionDeclaration) =
+        let name = typeDecl.SourceName
+        let typeArgs = typeDecl.TypeParameterDeclarationList
+        if isNull typeArgs then name else
+        let typeParams = typeArgs.TypeParameters
+        if Seq.isEmpty typeParams then name else
+
+        let typeArgsRepresentation =
+            typeParams
+            |> Seq.map _.NameIdentifier.GetTypeParameterOrSourceName()
+            |> String.concat ", "
+
+        match typeArgs with
+        | :? IPostfixTypeParameterDeclarationList ->  $"{name}<{typeArgsRepresentation}>"
+        | _ -> $"{typeArgsRepresentation} {name}"
+
+    static let addBinding (binding: IBindingLikeDeclaration) (context: SummarizerContext) =
+        let referencePat = binding.HeadPattern.As<IReferencePat>()
+        if isNull referencePat then () else
+
+        let typeRepr = referencePat.GetFcsSymbol() |> formatFcsSymbolType
+        let representation = $"val {referencePat.GetFcsSymbol().DisplayName}: {typeRepr}"
+        context.AddEntity(binding, representation)
```

Comment:
These tree lines are repeated several times, with the only difference being the representation string. Could you try extracting it?